### ISSUE TITLE
Proposal: Pluggable AuthN/AuthZ

### ISSUE THREAD

**Issue by etiennedi** [creativesoftwarefdn/weaviate#628]
## Business Background and Requirements
- we need flexibility because we assume every enterprise company has a slightly different Identity/Auth setup
- we want to be compatible with (and prefer) standards and best practices, but at the same time allow highly-customized solutions as well
- we assume different users have different needs for
  - Authentication: LDAP, OAuth2, simple basic Auth...
  - Authorization: smaller setups require something as simple as "every authenticated user can view, some people can write". More complex setup will require a very fine-graid auth "users of group foo can read `a.b.c` and `a.b.e`, but not `a.b.d`, but can write only `a.b.e`

## Proposed Solution

tl;dr:

* both Authentication (AuthN) and Authorization (AuthZ) are fully pluggable
* we'll start with very simple "plugins" like 
  * No Auth (for AuthN)
  * Basic Auth with fixed config (for AuthN)
  * Every Authenticated user is authorzied (for AuthZ)
  * etc.
* we will extend it with more elaborate plugins like OAuth2/OpenIDConnect (for AuthN) and Role-Based-Access (RBAC) for AuthZ later 
* there is a minimal interface between AuthN and AuthZ to reduce coupling
* overall inspired by Kubernetes, especially plugin "Role Based Access Control" 

### Architectural Components, Boundaries and Interfaces

![pluggable authn 2f authz overview 4](https://user-images.githubusercontent.com/8974479/49877981-e2146580-fe26-11e8-80a5-3b1325b5b365.png)

##### Incoming Request
Any request that comes in from the outside (wether it's REST or GraphQL) will go through both the Authentication and Authorization middleware.

##### Authentication Middleware (fixed, not pluggable)
* The first middleware to be called. It receives the whole request.
* It calls the Authentication plugin with  `request *http.Request`
* It receives `authenticated bool, authInfo AuthInfo`
* if it received `authenticated == false`, it aborts the request, returns a response with `HTTP Status 401`
* it received `authenticated == true`, it continues by calling the next middleware, the Authorization middleware with `authInfo AuthInfo`

##### Authentication Plugin
* has access to the full request (`*http.Request)
* decides based on configuration and the request whether the user is authenticated
* MUST extract a `username string`
* SHOULD extract a `group string`
* CAN extract `additionalData interface{}`
* MUST return `authenticated bool, authInfo{ username, group, additionalData}` 

##### Authorization Middleware (fixed, not pluggable)
* is called by the Authentication Middleware
* only if Authentication was successful
* It calls the Authorization plugin with `authInfo, resourcePath string, action string`
* It receives `authorized bool, err error`
* if it received `authorized == false`, it aborts the request, returns a response with `HTTP Status 403` and includes the error message, .e.g. `User 'JohnDoe@gmail.com' does not have permission to READ Things.City`
* it received `authorized == true`, it continues with the intended action.
 
##### Authorization Plugin
* has access to everything the Authentication Plugin extracted, i.e. `authInfo{ username, group, additionalData}`
* has access to a description of the desired resource and the desired action, i.e. `READ Local.Things.Cities`
MUST decide based on available info and configuration whether the user is authorized to perform the desired action

##### Original Request Handler
* unchanged, only called if both Authentication and Authorization were successful

##### Anonymous Access
If this plugin is activated every single request will be authenticated as user `anonymous` in group `anonymous`. Should be combined with an AuthZ plugin that gives `anonymous` some rights.

##### Basic Auth
If this plugin is activated every single request will be asked for a username and password. This will be validated against a configuration file. It will store plain-text passwords and is considered a simple tool to get started, but not a fully secure solution. Will resolve with the `username` and map this user to a preconfigured `group`. A configuration file for this plugin could look like
```
user,group,password
johndoe,admin,s3cr3tp4$$w0rd
```

##### OpenID connect
Expects a JWT to be sent as part of the request headers. Validates it against the token issuer specified in the configuration. Extracts subject (or other configurable field) as `username`, tries to extract a group if set in the token.

A possible configuration for this plugin could look like
```yaml
# openid-config.yml
issuer: http://my-identity-provider.com
audience: my-weaviate
usernameClaim: sub
groupClaim: group
```

##### Anonymous Plugin
A plugin with very few configuration options, can be set to either give users of the group `anonymous` full access or read-only access.

##### Simple Matrix (similar to RBAC, but more primitive)
A plugin that can map authenticated users or groups to allowed actions, but not configurable at runtime. Configuration could look something like

```yaml
# matrix-auth.yml
- username: johndoe@gmail.com
  resourcePath: ["Things", "City"]
  actions: ["read", "write"]
- group: anonymous
  resourcePath: ["Actions", "*"]
  actions: ["read"]
```
Explanation: 
- JohnDoe can read and write all `Things.City`, but cannot read or write any other Thing or Action
- Users in the group `anonymous` can read all Actions, but not write any. They can't read or write Things

##### RBAC
Same as Matrix, but configurable at runtime through API 

### Potential delivery Roadmap

1. Remove current Key-based system
1. Start with making the authentication pluggable, but ignore AuthZ. As part of this build the most simple authenticator, such as "Anonymous Access"
1. Add another simple authenticator, such as "Basic Auth"
1. Make Authorization pluggable, start with very simple authorizer, such as "Anonymous Access"
1. Slowly implement Matrix Plugin (can in turn happen in many iterations)
1. Replace Matrix Plugin with full RBAC solution that can be configured via API

**Comment by bobvanluijt** [comment]
Some thoughts @etiennedi,

1. Maybe it is smart to remove the key-based system asap. This allows people to use Weaviate asap.
2. I'm all for the proposed setup.
3. We have different authentication for P2P nodes.

**Comment by etiennedi** [comment]
Considering the major building blocks of this proposal accepted. Details can be ironed out down the line. Therefore closing this issue and opening two new issues to reflect Authentication and Authorization.
